src/group/libp2p_group_worker.erl

```diff
@@ -231,17 +232,40 @@ connecting(info, connect_retry_timeout, Data=#data{tid=TID,
         false ->
             Parent = self(),
             Pid = erlang:spawn_link(fun() ->
-                case libp2p_swarm:dial_framed_stream(TID, MAddr, Path, M, A) of
-                    {error, Error} ->
-                        Parent ! {connect_error, Error};
-                    {ok, StreamPid} ->
-                        Parent ! {assign_stream, StreamPid}
+                LocalAddr = libp2p_swarm:p2p_address(TID),
+                %% its possible that after exchanging identities two peers will be assigned each other as a target
+                %% and end up dialing each other at pretty much the same time
+                %% resulting in streams being assigned for the first dial, followed by a second set of streams for the second dial
+                %% there is/was then a coin toss ( see handle_assign_stream ) as to which streams win ( we dont want two streams between the same set of peers )
```

Comment:
I think it would be better to make the coin toss deterministic based on the identities of the peers involved, then they'd both make the same decision and we'd avoid getting an unlucky coin toss.

Reply:
@Vagabond So at the point of dialing ( as far as i can see ) we do not necessarily have the identity of the remote peer, for instance in the case of a new peer we have received over gossip from another connected peer ( which is why i am using the p2p addresses, as these are always available ).  In this scenario, these peers wont have a prior connect attempt, the connect will be performed as part of this initial dialing.

but i am wondering are we overthinking this...If the coin toss is made deterministic then its not a coin toss and if the decision is not based on anything related to the context of the existing or new stream then why use some arbitrary rule to make the decision ?  Why not hardcode the rule to always go with the new stream or stay with the old.  If there is no data or context to influence which option is "best" then why do we care other than both sides need to do the same thing.

One consideration we should keep in mind is, if we decide to drop the existing and go with the new stream then we risk loosing some in flight gossiped msgs.  If we dont care about msg loss here ( i mean a dropped block will be gossiped again, we wont miss a dropped gossiped peer ? and relcast msgs are acked iirc anyway and thus will be resent ) then why not always just accept the new stream on both sides ?

Reply:
You only need to flip a coin if you get a simultaneous dial. Using the local p2p address and the remote p2p address to make the coin flip deterministic would simply be flipping the same coin, rather than both sides flipping coins independently.

Reply:
yes, the coin toss does only take place when there is a simultaneous dial ( which is how it was before any changes in this PR )...but flipping the coins independently is where the problem comes in...if both sides flip and get different results they each close a different stream resulting in both old and new streams being closed

Maybe i am not seeing it but why would this be desirable behaviour ?

Reply:
Right, my point is to replace the independent coin flip with a common one. Instead of each side doing rand(2) after a simultaneous dial, they'd instead look at their local identity and the remote entity they're dialing and use that *common* information to decide if they should *both* pick heads or tails.

Reply:
gotcha, but keep in mind that dialing can occur before we know the remote peer identity...all we have in some cases is the peerbook entry.  The identities will be swapped as part of the connect performed after dialing is triggered.

But on a general point, what benefit does this coinflip provide ( even if it is a common flip ) ?  It is detached from any context relating to one stream or the other.  Why would we not just say always go with the latest stream or always stick with the existing steam ( on both sides ).  I dont see the value in varying that decision based on the coin flip.

Reply:
Things like 'latest' or 'existing' are not common on both sides, those are relative values. Things like clock skew or network round trip time can really give both sides a very different view. There needs to be some objective, independent way to break the tie.

When I say Identity, I mean simply the p2p address. That has plenty of entropy and is unique so we can use that to make the decision on which end to chooose.

For example, you could, for a simultaneous connection from a->b and b->a look at if a is greater than b, if it is, you pick the a->b connection (where a is the dialer and b is the dialee) otherwise pick the b->a connedtion, both sides can make this decision independently and come to the same conclusion.

Reply:
hmm i am going to think about your last statement further because from my perspective  latest and existing are/were common on both sides.  'existing' is the  currently connected stream from a->b whilst the 'latest' is any new stream which has been established after the 'existing' one connected

but...what may be factoring in here is that the group workers spawn off dialing which i have been thinking is odd given the process is an async worker with no real sync APIs.  Spawning off the dialing however is likely going to exasperate the potential for a worker to get assigned into an undesirable state by dealing with other msgs whist an existing dial is ongoing....not sure what exact implications might be but it smells bad and may be related to this overall problem....

Reply:
no, in the case of a simultaneous dial you have the following:

a is dialing b
b is dialing a

from a's perspective a->b is the existing connection, and the older one (because it started first from its perspective)
from b's perspetive, b->a is the existing connection, and the older one (because it started first from its perspective)

When a->b connects, b sees it as 'newer' than b->a, but a sees it as older than b->a.

If we simply prefer existing connections, each side would pick the connection with themselves as the dialer, and close the other one, thus both would get closed.

I don't think the async thing is a problem, the problem is that the far end (which may be on the other side of the world) has a different perspective than we do, and is operating independently.

Reply:
@Vagabond i see the potential for what you describe.  That said it is not what I was observing with the tests.  With the test runs on buildkite it was simpler.  

a is dialing b
b is dialing a

a's dial completes ( both sides see this as the first/existing connection )
b's dial complete ( both sides see this as the second/new connection )

Both sides toss the coin and get opposite sides resulting in both connections closing.  
If both sides get the same side, all is well.

I'll rework this PR to allow both dials and handle in the coin toss.  Took me a while to see this, thanks for discussing.

Reply:
Ah yes, on localhost there's usually not enough network latency for the dials to overlap in time. If there's a clear winner we should prefer that and only resort to coin tossing if absolutely necessary.

That said, you don't know what your peer thinks is going on, so if you make a unilateral decision based on your perspective you're likely to be wrong if there's significant latency involved.